app/Actions/Album/Archive.php

```diff
@@ -32,18 +36,24 @@ class Archive extends Action
 		'<', '>', ':', '"', '/', '\\', '|', '?', '*',
 	];
 
+	protected int $deflateLevel = -1;
+
 	/**
 	 * @param Collection<AbstractAlbum> $albums
 	 *
 	 * @return StreamedResponse
 	 *
 	 * @throws FrameworkException
+	 * @throws ConfigurationKeyMissingException
 	 */
 	public function do(Collection $albums): StreamedResponse
 	{
+		$this->deflateLevel = Configs::getValueAsInt('zip_deflate_level');
+
 		$responseGenerator = function () use ($albums) {
-			$options = new \ZipStream\Option\Archive();
+			$options = new ZipArchiveOption();
 			$options->setEnableZip64(Configs::getValueAsBool('zip64'));
+			$options->setZeroHeader(true);
```

Comment:
The documentation on this is cryptic. Is this a speed optimization?

Reply:
Yes and no. I believe you are referring to this sentence

> Enable streaming files with single read where general purpose bit 3 indicates local file header contain zero values in crc and size fields, these appear only after file contents in data descriptor block. Default is false. Set to true if your input stream is remote (used with addFileFromStream()).

from [ZipStream-PHP: Available-options]. I am not even sure if the first sentence is actually a sentence :D

I am not 100% sure what id does, but I did some step debugging through the code and https://en.wikipedia.org/wiki/Deflate gave me some clues. My assumption is also backed up by looking at some ZIP archives in the hex editor.

I believe the documentation should have been as follows

> The DEFLATE algorithm allows to use a Huffman encoding which is globally optimized and constructed for the entire archive or individual, local Huffman encoding for each processing block. On a technical level the former type of blocks inside an archive are indicated by a `x01`-header, the latter type is indicated by a `x10-header (cp. [Wikipedia: Deflate](https://en.wikipedia.org/wiki/Deflate)).
> In order to get the next to optimal compression ratio the DEFLATE algorithm needs to read the entire payload three times (across file boundaries). This allows to construct a near to optimal Huffman encoding for the entire archive on a global level. > However a 3-pass scan is prohibitive for large archives and does not work for file streams which are non-seekable, like remote file streams.

But this contains a lot of guess work here. I have no clue what the documentation wants to say about "contain zero values in crc and size fields". But I can confirm that without `setZeroHeader(true)` the library tries to read every file three times.

Reply:
It sounds like they set the crc and size to `0` in the header but also set some flag to indicate that the actual values will be provided after the payload (the compressed file contents). Otherwise an extra pass over the file contents would be needed just to calculate the crc (and size, if unknown) before the compression can even start.